cpp/src/arrow/dataset/scanner.h

```diff
@@ -318,6 +337,28 @@ class ARROW_DS_EXPORT SyncScanner : public Scanner {
   std::shared_ptr<Fragment> fragment_;
 };
 
+class ARROW_DS_EXPORT AsyncScanner : public Scanner,
```

Comment:
Is this worth declaring in the header? We could keep it in the .cc file so that people don't instantiate it directly.

Reply:
Ah, that's a good idea.  I can move both the scanners to `.cc`.

Reply:
Done.